From 99744a5bd62c9184f872301d00ee8f8ea3b1fbdd Mon Sep 17 00:00:00 2001
From: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Date: Thu, 11 Oct 2018 15:54:25 +0530
Subject: [PATCH 0126/1852] ASoC: xlnx: support multiple sampling rates for
 HDMI

commit 37fe7003143f04fcfb85b24060ef11a0aade700a from
https://github.com/Xilinx/linux-xlnx.git

This adds support for HDMI playback at various supported sampling rates.

Signed-off-by: Maruthi Srinivas Bayyavarapu <maruthi.srinivas.bayyavarapu@xilinx.com>
Reviewed-by: Sandip Kothari <sandipk@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 sound/soc/xilinx/xlnx_pl_snd_card.c | 32 +++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/sound/soc/xilinx/xlnx_pl_snd_card.c b/sound/soc/xilinx/xlnx_pl_snd_card.c
index 3786c7d79656..0d0411ad83f2 100644
--- a/sound/soc/xilinx/xlnx_pl_snd_card.c
+++ b/sound/soc/xilinx/xlnx_pl_snd_card.c
@@ -41,6 +41,33 @@ static struct snd_soc_card xlnx_card = {
 	.owner = THIS_MODULE,
 };
 
+static int xlnx_hdmi_card_hw_params(struct snd_pcm_substream *substream,
+				    struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct pl_card_data *prv = snd_soc_card_get_drvdata(rtd->card);
+	u32 sample_rate = params_rate(params);
+
+	switch (sample_rate) {
+	case 32000:
+	case 44100:
+	case 48000:
+	case 96000:
+	case 176400:
+	case 192000:
+		prv->mclk_ratio = 768;
+		break;
+	case 88200:
+		prv->mclk_ratio = 192;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	prv->mclk_val = prv->mclk_ratio * sample_rate;
+	return clk_set_rate(prv->mclk, prv->mclk_val);
+}
+
 static int xlnx_i2s_card_hw_params(struct snd_pcm_substream *substream,
 				   struct snd_pcm_hw_params *params)
 {
@@ -112,6 +139,10 @@ static const struct snd_soc_ops xlnx_i2s_card_ops = {
 	.hw_params = xlnx_i2s_card_hw_params,
 };
 
+static const struct snd_soc_ops xlnx_hdmi_card_ops = {
+	.hw_params = xlnx_hdmi_card_hw_params,
+};
+
 SND_SOC_DAILINK_DEFS(xlnx_i2s,
 		     DAILINK_COMP_ARRAY(COMP_CPU("xilinx-i2s")),
 		     DAILINK_COMP_ARRAY(COMP_DUMMY()),
@@ -154,6 +185,7 @@ static struct snd_soc_dai_link xlnx_snd_dai[][XLNX_MAX_PATHS] = {
 		{
 			.name = "xilinx-hdmi-playback",
 			SND_SOC_DAILINK_REG(xlnx_hdmi_tx),
+			.ops = &xlnx_hdmi_card_ops,
 		},
 		{
 			.name = "xilinx-hdmi-capture",
-- 
2.31.1

